contracts/libraries/CronUtility.sol

```diff
@@ -0,0 +1,476 @@
+   * @param minInterval the minimum interval of the field (usually 1)
+   * @param min the minimum value a field can have (usually 1 or 0)
+   * @param max the maximum value a field can have (ex minute = 59, hour = 23)
+   */
+  function validateField(
+    Field memory field,
+    string memory fieldName,
+    uint8 minInterval,
+    uint8 min,
+    uint8 max
+  ) private pure {
+    if (field.fieldType == WILD) {
+      return;
+    } else if (field.fieldType == EXACT) {
+      if (field.singleValue < min || field.singleValue > max) {
+        string memory reason = string(
+          bytes.concat(
+            "value must be >=,",
+            uintToBString(min),
+            " and <=",
+            uintToBString(max)
+          )
+        );
+        revert invalidField(fieldName, reason);
+      }
+    } else if (field.fieldType == INTERVAL) {
+      if (field.interval < minInterval || field.interval > max) {
+        string memory reason = string(
+          bytes.concat(
+            "inverval must be */(",
+            uintToBString(minInterval),
+            "-",
+            uintToBString(max),
+            ")"
+          )
+        );
+        revert invalidField(fieldName, reason);
+      }
+    } else if (field.fieldType == RANGE) {
+      if (field.rangeEnd > max) {
+        string memory reason = string(
+          bytes.concat(
+            "inverval must be within ",
+            uintToBString(min),
+            "-",
+            uintToBString(max)
+          )
+        );
+        revert invalidField(fieldName, reason);
+      }
+    } else if (field.fieldType == LIST) {
+      string memory reason = string(
+        bytes.concat(
+          "items in list must be within ",
+          uintToBString(min),
+          "-",
+          uintToBString(max)
+        )
+      );
+      uint8 listItem;
+      for (uint256 idx = 0; idx < field.listLength; idx++) {
+        listItem = field.list[idx];
+        if (listItem < min || listItem > max) {
+          revert invalidField(fieldName, reason);
+        }
+      }
+    } else {
+      revert unknownFieldType();
+    }
+  }
+
+  // TYPE CONVERSIONS
+
+  /**
+   * @notice toSpec converts a cron string to a spec struct. This is gas-intensive
+   * and therefore should only be called off-chain.
+   * @param cronString the cron string
+   * @return the spec struct
+   */
+  function toSpec(string memory cronString) public pure returns (Spec memory) {
+    strings.slice memory space = strings.toSlice(" ");
+    strings.slice memory cronSlice = strings.toSlice(cronString);
+    if (cronSlice.count(space) != 4) {
+      revert invalidSpec("4 spaces required");
+    }
+    strings.slice memory minuteSlice = cronSlice.split(space);
+    strings.slice memory hourSlice = cronSlice.split(space);
+    strings.slice memory daySlice = cronSlice.split(space);
+    strings.slice memory monthSlice = cronSlice.split(space);
+    // DEV: dayOfWeekSlide = cronSlice
```

Comment:
the `cronSlice` is always reassigned to the rest of the string after running `split()`. After the final split, the `cronSlice` is equal to the dayOfWeek. Assigning it would be an unnecessary memory allocation, however.

aka...
```solidity
strings.slice memory dayOfWeekSlice = cronSlice
```

(oops, I'll fix the typo!)

Reply:
Maybe this will be less confusing...

```solidity
// DEV: save an extra allocation and just use cronSlice as dayOfWeekSlice
// strings.slice memory dayOfWeekSlice = cronSlice;
```

Reply:
Possibly more clear to just assign a new variable, but I don't feel strongly one way or the other.